algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py

```diff
@@ -0,0 +1,382 @@
+    self.dtype = dtype
+
+    num_patches = (self.image_height // self.patch_size[0]) * \
+                  (self.image_width // self.patch_size[1])
+    if self.posemb == 'learn':
+      self.pos_embedding = nn.Parameter(torch.randn(1, num_patches, width))
+      self.pos_embedding.data.normal_(std=1 / math.sqrt(self.width))
+
+    if self.pool_type == 'tok':
+      self.cls = nn.Parameter(torch.randn(1, 1, width)).type(self.dtype)
+      self.cls.data.zero_()
+
+    if self.rep_size:
+      rep_size = self.width if self.rep_size is True else self.rep_size  # pylint: disable=g-bool-id-comparison
+      self.pre_logits = nn.Linear(self.width, rep_size)
+      init_weights(self.pre_logits)
+
+    self.embedding = nn.Conv2d(
+        self.channels,
+        self.width,
+        self.patch_size,
+        stride=self.patch_size,
+        padding='valid')
+    init_weights(self.embedding)
+    self.dropout = nn.Dropout(p=dropout)
+
+    self.encoder = Encoder(
+        depth=self.depth,
+        width=self.width,
+        mlp_dim=self.mlp_dim,
+        num_heads=self.num_heads,
+        dropout=dropout)
+
+    if self.num_classes:
+      self.head = nn.Linear(self.width, self.num_classes)
+      init_weights(self.head)
+    if self.head_zeroinit:
+      self.head.weight.data.zero_()
+
+  def get_posemb(self, x: Tensor) -> Tensor:
+    if self.posemb == 'learn':
+      return self.pos_embedding.type(self.dtype)
+    elif self.posemb == 'sincos2d':
+      return posemb_sincos_2d(x).type(self.dtype)
+    else:
+      raise ValueError(f'Unknown posemb type: {self.posemb}')
+
+  def forward(self, x: Tensor) -> Tensor:
+    out = {}
+
+    # Patch extraction
+    x = out['stem'] = self.embedding(x)
+
+    # Add posemb before adding extra token.
+    n, c, h, w = x.shape
+    pes = self.get_posemb(x)
+
+    # Reshape to match Jax's ViT implementation.
+    x = torch.transpose(torch.reshape(x, (n, c, h * w)), 1, 2)
+    x = out['with_posemb'] = x + pes
+
+    if self.pool_type == 'tok':
+      x = torch.cat((torch.tile(self.cls, [n, 1, 1]), x), dim=1)
+
+    n, l, c = x.shape  # pylint: disable=unused-variable
+    x = self.dropout(x)
+
+    x, out['encoder'] = self.encoder(x)
+    encoded = out['encoded'] = x
+
+    if self.pool_type == 'gap':
+      x = out['head_input'] = torch.mean(x, dim=1)
+    elif self.pool_type == '0':
+      x = out['head_input'] = x[:, 0]
+    elif self.pool_type == 'tok':
+      x = out['head_input'] = x[:, 0]
+      encoded = encoded[:, 1:]
+    else:
+      raise ValueError(f'Unknown pool type: "{self.pool_type}"')
+
+    x_2d = torch.reshape(encoded, [n, h, w, -1])
+
+    if self.rep_size:
+      hid = self.pre_logits(x_2d)
+      x_2d = torch.tanh(hid)
+      x = torch.tanh(hid)
+
+    out['pre_logits_2d'] = x_2d
+    out['pre_logits'] = x
+
+    if self.num_classes:
+      x_2d = out['logits_2d'] = self.head(x_2d)
+      x = out['logits'] = self.head(x)
+    return x, out
+
+
+def decode_variant(variant):
```

Comment:
given this is the same as the jax version, could we move this into a common file?

Reply:
Thank you! I moved this function into the shared `workload.py` file, which should be addressed in the commit [f6c542a](https://github.com/mlcommons/algorithmic-efficiency/pull/82/commits/f6c542a78d295d7a9c00dde8bd15fbde70f4ccc7).